src/core/Akka.Streams/Implementation/IO/OutputStreamSourceStage.cs

```diff
@@ -100,10 +106,23 @@ private sealed class Logic : GraphStageLogic, IStageWithCallback
                 });
                 _upstreamCallback =
                     GetAsyncCallback<Tuple<IAdapterToStageMessage, TaskCompletionSource<NotUsed>>>(OnAsyncMessage);
-                _pullTask = new OnPullRunnable(downstreamCallback, dataQueue);
+                _pullTask = new OnPullRunnable(downstreamCallback, dataQueue, ref _blockingThread);
                 SetHandler(_stage._out, onPull: OnPull, onDownstreamFinish: OnDownstreamFinish);
             }
 
+            public override void PreStart()
+            {
+                _dispatcher = ActorMaterializer.Downcast(Materializer).System.Dispatchers.Lookup(_dispatcherId);
+                base.PreStart();
+            }
+
+            public override void PostStop()
+            {
+                // interrupt any pending blocking take
+                _blockingThread?.Interrupt();
```

Comment:
Thread.Interrupt dosn't work on Mono and won't work on NetCore. Right now we have an error
```
[20:41:21] :		 [docker] [WARNING][10/01/2016 20:41:21][Thread 0001][ActorSystem(FlowSelectAsyncUnorderedSpec-190)] NewtonSoftJsonSerializer has been detected as a default serializer. It will be obsoleted in Akka.NET starting from version 1.5 in the favor of Wire (for more info visit: http://getakka.net/docs/Serialization#how-to-setup-wire-as-default-serializer ). If you want to suppress this message set HOCON `akka.suppress-json-serializer-warning` config flag to on.
[20:41:21] :			 [Test collection for Akka.Streams.Tests.Dsl.FlowSelectAsyncUnorderedSpec (1)] Akka.Streams.Tests.Dsl.FlowSelectAsyncUnorderedSpec.A_Flow_with_SelectAsyncUnordered_must_not_run_more_futures_than_configured (11s)
[20:41:21] :		 [docker] System.Threading.ThreadInterruptedException: 
[20:41:22]W:		 [docker] Running build failed.
[20:41:22]W:		 [docker] Error:
[20:41:22]W:		 [docker] System.Exception: xUnit2 reported an error (Error Code 1)
[20:41:22]W:		 [docker]   at Fake.Testing.XUnit2+ResultHandling+failBuildWithMessage@199-3.Invoke (System.String message) <0x41136240 + 0x0001f> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Microsoft.FSharp.Core.OptionModule.Iterate[T] (Microsoft.FSharp.Core.FSharpFunc`2 action, Microsoft.FSharp.Core.FSharpOption`1 option) <0x411359a0 + 0x0002e> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Fake.Testing.XUnit2+ResultHandling+failBuildIfXUnitReportedError@203-2.Invoke (Microsoft.FSharp.Core.FSharpOption`1 option) <0x41135960 + 0x00023> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Fake.Testing.XUnit2+ResultHandling+failBuildIfXUnitReportedError@202-3.Invoke (Int32 x) <0x41135670 + 0x00034> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Fake.Testing.XUnit2.xUnit2 (Microsoft.FSharp.Core.FSharpFunc`2 setParams, IEnumerable`1 assemblies) <0x41128630 + 0x0016d> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at FSI_0005.Build+runSingleAssembly@279.Invoke (System.String assembly) <0x41128430 + 0x0011f> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Microsoft.FSharp.Collections.SeqModule.Iterate[T] (Microsoft.FSharp.Core.FSharpFunc`2 action, IEnumerable`1 source) <0x410881b0 + 0x000ce> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at FSI_0005.Build+clo@260-27.Invoke (Microsoft.FSharp.Core.Unit _arg2) <0x41125380 + 0x002ef> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Fake.TargetHelper+targetFromTemplate@195[a].Invoke (Microsoft.FSharp.Core.Unit unitVar0) <0x411246e0 + 0x00023> in <filename unknown>:0 
[20:41:22]W:		 [docker]   at Fake.TargetHelper.runSingleTarget (Fake.TargetTemplate`1 target) <0x41119b10 + 0x000ca> in <filename unknown>:0 
```